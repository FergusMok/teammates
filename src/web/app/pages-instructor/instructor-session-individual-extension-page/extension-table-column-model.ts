export interface StudentExtensionTableColumnModel {
    sectionName: string;
    teamName: string;
    name: string;
    email: string;
    extensionDeadline: number;
    hasExtension: boolean;
    selected: boolean;
  }

  export interface InstructorExtensionTableColumnModel {
    institute?: string;
    name: string;
    email: string;
    extensionDeadline: number;
    hasExtension: boolean;
    selected: boolean;
  }
