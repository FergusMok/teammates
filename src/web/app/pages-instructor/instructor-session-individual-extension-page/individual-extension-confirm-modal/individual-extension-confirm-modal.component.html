<div class="modal-header alert-primary">
  <h5 class="modal-title">Confirm extension of feedback session</h5>
  <button type="button" class="close" (click)="activeModal.dismiss()">
    <i class="fas fa-times"></i>
  </button>
</div>

<div class="modal-body">
  The new deadline of <b>{{ extensionTimestamp | date:'d MMM yyyy hh:mm' }}</b> will be applied
  to the following <b>{{studentsToExtend.length}} students</b> upon confirmation:
  <div class="container ml-1">
</div>

<div class="modal-body">
  <div class="table-responsive">
    <table id="student-list-table" class="table table-bordered">
      <thead>
        <tr>
          <th class="sortable-header" (click)="sortCoursesBy(SortBy.SECTION_NAME)">
            Section
            <span class="fa-stack">
              <i class="fas fa-sort"></i>
              <i *ngIf="sortBy === SortBy.SECTION_NAME && sortOrder === SortOrder.DESC" class="fas fa-sort-down"></i>
              <i *ngIf="sortBy === SortBy.SECTION_NAME && sortOrder === SortOrder.ASC" class="fas fa-sort-up"></i>
            </span>
          </th>
          <th class="sortable-header" (click)="sortCoursesBy(SortBy.TEAM_NAME)">
            Team
            <span class="fa-stack">
              <i class="fas fa-sort"></i>
              <i *ngIf="sortBy === SortBy.TEAM_NAME && sortOrder === SortOrder.DESC" class="fas fa-sort-down"></i>
              <i *ngIf="sortBy === SortBy.TEAM_NAME && sortOrder === SortOrder.ASC" class="fas fa-sort-up"></i>
            </span>
          </th>
          <th class="sortable-header" (click)="sortCoursesBy(SortBy.RESPONDENT_NAME)">
            Name
            <span class="fa-stack">
              <i class="fas fa-sort"></i>
              <i
                *ngIf="sortBy === SortBy.RESPONDENT_NAME && sortOrder === SortOrder.DESC"
                class="fas fa-sort-down"
              ></i>
              <i *ngIf="sortBy === SortBy.RESPONDENT_NAME && sortOrder === SortOrder.ASC" class="fas fa-sort-up"></i>
            </span>
          </th>
          <th class="sortable-header" (click)="sortCoursesBy(SortBy.RESPONDENT_EMAIL)">
            Email
            <span class="fa-stack">
              <i class="fas fa-sort"></i>
              <i
                *ngIf="sortBy === SortBy.RESPONDENT_EMAIL && sortOrder === SortOrder.DESC"
                class="fas fa-sort-down"
              ></i>
              <i *ngIf="sortBy === SortBy.RESPONDENT_EMAIL && sortOrder === SortOrder.ASC" class="fas fa-sort-up"></i>
            </span>
          </th>
          <th class="sortable-header" (click)="sortCoursesBy(SortBy.SESSION_END_DATE)">
            Original Deadline
            <span class="fa-stack">
              <i class="fas fa-sort"></i>
              <i
                *ngIf="sortBy === SortBy.SESSION_END_DATE && sortOrder === SortOrder.DESC"
                class="fas fa-sort-down"
              ></i>
              <i *ngIf="sortBy === SortBy.SESSION_END_DATE && sortOrder === SortOrder.ASC" class="fas fa-sort-up"></i>
            </span>
          </th>
        </tr>
      </thead>
      <tr *ngFor="let student of studentsToExtend; index as i">
        <td>{{ student.sectionName }}</td>
        <td>{{ student.teamName }}</td>
        <td>{{ student.studentName }}</td>
        <td>{{ student.studentEmail }}</td>
        <td>{{ student.studentExtensionDeadline | date:'d MMM yyyy hh:mm' }}</td>
      </tr>
    </table>
  </div>
</div>

  <div class="form-check">
    <input class="form-check-input" type="checkbox" [(ngModel)]="isNotifyStudents" id="flexCheckChecked">
    <label class="form-check-label" for="flexCheckChecked">
      Notify students by email about this extension
    </label>
  </div>

</div>

<br>

<div class="modal-footer">
  <button type="button" class="btn btn-light" (click)="activeModal.dismiss()">No, Cancel the Operation</button>
  <button type="button" class="btn btn-success modal-btn-ok" (click)="onExtend()">Confirm</button>
</div>
