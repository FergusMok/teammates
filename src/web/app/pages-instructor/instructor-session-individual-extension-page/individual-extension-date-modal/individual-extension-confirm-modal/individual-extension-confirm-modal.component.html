<div class="modal-header alert-primary">
  <h5 class="modal-title">Confirm extension of feedback session</h5>
  <button type="button" class="close" (click)="activeModal.dismiss()">
    <i class="fas fa-times"></i>
  </button>
</div>

<div class="modal-body">
  The following changes to the individual deadlines for <b>11 students</b> will be saved upon confirmation:
  <div class="container ml-1">
</div>
<div class="modal-body">
  <div class="table-responsive">
    <table id="student-list-table" class="table table-bordered">
      <tr class="background-color-medium-gray">
        <th  class="clickable">
          No.
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.SECTION_NAME && studentListInfoTableSortOrder === SortOrder.DESC" class="fas fa-sort-down"></i> -->
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.SECTION_NAME && studentListInfoTableSortOrder === SortOrder.ASC" class="fas fa-sort-up"></i> -->
          </span>
        </th>
        <th  class="clickable">
          Section
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.SECTION_NAME && studentListInfoTableSortOrder === SortOrder.DESC" class="fas fa-sort-down"></i> -->
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.SECTION_NAME && studentListInfoTableSortOrder === SortOrder.ASC" class="fas fa-sort-up"></i> -->
          </span>
        </th>
        <th class="clickable">
          Team
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.TEAM_NAME && studentListInfoTableSortOrder === SortOrder.DESC" class="fas fa-sort-down"></i> -->
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.TEAM_NAME && studentListInfoTableSortOrder === SortOrder.ASC" class="fas fa-sort-up"></i> -->
          </span>
        </th>
        <th class="clickable">
          Student Name
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.RESPONDENT_NAME && studentListInfoTableSortOrder === SortOrder.DESC" class="fas fa-sort-down"></i> -->
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.RESPONDENT_NAME && studentListInfoTableSortOrder === SortOrder.ASC" class="fas fa-sort-up"></i> -->
          </span>
        </th>
        <th class="clickable">
          Email
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.RESPONDENT_EMAIL && studentListInfoTableSortOrder === SortOrder.DESC" class="fas fa-sort-down"></i> -->
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.RESPONDENT_EMAIL && studentListInfoTableSortOrder === SortOrder.ASC" class="fas fa-sort-up"></i> -->
          </span>
        </th>
        <th class="clickable">
          Original Deadline
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.RESPONDENT_EMAIL && studentListInfoTableSortOrder === SortOrder.DESC" class="fas fa-sort-down"></i> -->
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.RESPONDENT_EMAIL && studentListInfoTableSortOrder === SortOrder.ASC" class="fas fa-sort-up"></i> -->
          </span>
        </th>
        <th class="clickable">
          New Deadline
          <span class="fa-stack">
            <i class="fas fa-sort"></i>
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.RESPONDENT_EMAIL && studentListInfoTableSortOrder === SortOrder.DESC" class="fas fa-sort-down"></i> -->
            <!-- <i *ngIf="studentListInfoTableSortBy === SortBy.RESPONDENT_EMAIL && studentListInfoTableSortOrder === SortOrder.ASC" class="fas fa-sort-up"></i> -->
          </span>
        </th>
      </tr>
      <tr>
        <td>
          1
        </td>
        <td>
          Section 1
        </td>
        <td>
          Team 1
        </td>
        <td>
          Student 1
        </td>
        <td>
          student-1@gmail.com
        </td>
        <td>
          1st March 2021, 23:59H
        </td>
        <td>
          8th March 2021, 23:59H
        </td>
      </tr>
      <tr >
        <td>
          2
        </td>
        <td>
          Section 1
        </td>
        <td>
          Team 1
        </td>
        <td>
          Student 2
        </td>
        <td>
          student-2@gmail.com
        </td>
        <td>
          1st March 2021, 23:59H
        </td>
        <td>
          8th March 2021, 23:59H
        </td>
      </tr>
      <tr>
        <td>
          3
        </td>
        <td>
          Section 1
        </td>
        <td>
          Team 1
        </td>
        <td>
          Student 3
        </td>
        <td>
          student-3@gmail.com
        </td>
        <td>
          1st March 2021, 23:59H
        </td>
        <td>
          8th March 2021, 23:59H
        </td>
      </tr>
      <tr>
              <td>
          4
        </td>
        <td>
          Section 1
        </td>
        <td>
          Team 1
        </td>
        <td>
          Student 4
        </td>
        <td>
          student-4@gmail.com
        </td>
        <td>
          1st March 2021, 23:59H
        </td>
        <td>
          8th March 2021, 23:59H
        </td>
      </tr>
      <tr>
              <td>
          5
        </td>

        <td>
          Section 1
        </td>
        <td>
          Team 1
        </td>
        <td>
          Student 5
        </td>
        <td>
          student-5@gmail.com
        </td>
        <td>
          1st March 2021, 23:59H
        </td>
        <td>
          8th March 2021, 23:59H
        </td>
      </tr>
      <tr>
              <td>
          6
        </td>
        <td>
          Section 1
        </td>
        <td>
          Team 1
        </td>
        <td>
          Student 6
        </td>
        <td>
          student-6@gmail.com
        </td>
        <td>
          1st March 2021, 23:59H
        </td>
        <td>
          8th March 2021, 23:59H
        </td>
      </tr>
      <tr>
              <td>
          7
        </td>
        <td>
          Section 1
        </td>
        <td>
          Team 1
        </td>
        <td>
          Student 7
        </td>
        <td>
          student-7@gmail.com
        </td>
        <td>
          1st March 2021, 23:59H
        </td>
        <td>
          8th March 2021, 23:59H
        </td>
      </tr>
      <tr>
              <td>
          8
        </td>
        <td>
          Section 1
        </td>
        <td>
          Team 1
        </td>
        <td>
          Student 8
        </td>
        <td>
          student-8@gmail.com
        </td>
        <td>
          1st March 2021, 23:59H
        </td>
        <td>
          8th March 2021, 23:59H
        </td>
      </tr>
      <tr>
              <td>
          9
        </td>
        <td>
          Section 1
        </td>
        <td>
          Team 1
        </td>
        <td>
          Student 9
        </td>
        <td>
          student-9@gmail.com
        </td>
        <td>
          1st March 2021, 23:59H
        </td>
        <td>
          8th March 2021, 23:59H
        </td>
      </tr>
      <tr>
              <td>
          10
        </td>
        <td>
          Section 1
        </td>
        <td>
          Team 1
        </td>
        <td>
          Student 10
        </td>
        <td>
          student-10@gmail.com
        </td>
        <td>
          1st March 2021, 23:59H
        </td>
        <td>
          8th March 2021, 23:59H
        </td>
      </tr>
      <tr>
              <td>
          11
        </td>
        <td>
          Section 1
        </td>
        <td>
          Team 1
        </td>
        <td>
          Student 11
        </td>
        <td>
          student-11@gmail.com
        </td>
        <td>
          1st March 2021, 23:59H
        </td>
        <td>
          8th March 2021, 23:59H
        </td>
      </tr>
    </table>
  </div>
</div>

  <div class="form-check">
    <input class="form-check-input" type="checkbox" value="" id="flexCheckChecked" checked>
    <label class="form-check-label" for="flexCheckChecked">
      Notify students by email about this extension
    </label>
  </div>

</div>

<br>

<div class="modal-footer">
  <button type="button" class="btn btn-light" (click)="activeModal.dismiss()">No, Cancel the Operation</button>
  <button type="button" class="btn btn-success modal-btn-ok" (click)="activeModal.close()">Confirm</button>
</div>
