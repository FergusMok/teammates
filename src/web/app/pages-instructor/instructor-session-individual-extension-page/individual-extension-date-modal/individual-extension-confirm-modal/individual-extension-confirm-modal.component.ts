import { Component, Input, OnInit } from '@angular/core';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';

/**
 * Modal to confirm permanent deletion of a feedback session.
 */
@Component({
  selector: 'individual-extension-confirm-modal',
  templateUrl: './individual-extension-confirm-modal.component.html',
  styleUrls: ['./individual-extension-confirm-modal.component.scss'],
})
export class IndividualExtensionConfirmModalComponent implements OnInit {

  @Input()
  courseId: string = '';

  @Input()
  numberOfStudentsToExtend: number = 0;

  @Input()
  feedbackSessionName: string = '';

  constructor(public activeModal: NgbActiveModal) {}

  ngOnInit(): void {
  }

  setA(a: boolean) {
    console.log(a);
  }
  
  onExtend() { 
    this.activeModal.close()
  }
  
  onDelete()
  {
  }

}