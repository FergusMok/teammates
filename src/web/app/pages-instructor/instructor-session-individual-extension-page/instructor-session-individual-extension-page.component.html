<div *tmIsLoading="isLoadingFeedbackSession">
  <div class="card border-primary margin-top-20px">
    <div class="card-body">
      <div class="row text-center">
        <div class="col-md-2 text-md-right font-bold">Course ID</div>
        <div id="time-zone" class="col-md-3 text-md-left">
          {{ courseId }}
        </div>
      </div>
      <div class="row text-center">
        <div class="col-md-2 text-md-right font-bold">Time Zone</div>
        <div id="time-zone" class="col-md-3 text-md-left">
          {{ feedbackSessionTimeZone }}
        </div>
      </div>
      <div class="row text-center">
        <div class="col-md-2 text-md-right font-bold">Course Name</div>
        <div id="course-name" class="col-md-10 text-md-left">
          {{ courseName }}
        </div>
      </div>
      <div class="row text-center">
        <div class="col-md-2 text-md-right font-bold">Session Name</div>
        <div id="course-name" class="col-md-10 text-md-left">
          {{ feedbackSessionName }}
        </div>
      </div>
      <div class="row text-center">
        <div class="col-md-2 text-md-right font-bold">Original Deadline</div>
        <div id="course-name" class="col-md-10 text-md-left">
          {{ feedbackSessionEndingTime | date:'d MMM yyyy hh:mm' }}
        </div>
      </div>
    </div>
  </div>
  <div class="container ml-1">
    <div class="align-row">
      <div>
        <button type="button" id="extend-btn" class="btn btn-success" (click)="onExtend()" [disabled]="!hasSelectedStudents()">Extend</button>
      </div>
      <div>
        <button type="button" id="delete-btn" class="btn btn-danger" (click)="onDelete()" [disabled]="!hasSelectedStudents()">Delete</button>
      </div>
    </div>
    <small
      >Select students from list below and click apply or remove to change their feedback session deadline
      extension.</small
    >
  </div>
</div>

<div *tmIsLoading="isLoadingAllStudents">
  <div class="modal-body">
    <div class="table-responsive">
      <table id="student-list-table" class="table table-bordered">
        <thead>
          <tr>
            <th (click)="selectAllStudents()">
              <label><input type="checkbox" id="select-all-btn" [checked]="isAllSelected" /> Select all</label>
            </th>
            <th class="sortable-header" (click)="sortCoursesBy(SortBy.SECTION_NAME)">
              Section
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortBy === SortBy.SECTION_NAME && sortOrder === SortOrder.DESC" class="fas fa-sort-down"></i>
                <i *ngIf="sortBy === SortBy.SECTION_NAME && sortOrder === SortOrder.ASC" class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortCoursesBy(SortBy.TEAM_NAME)">
              Team
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortBy === SortBy.TEAM_NAME && sortOrder === SortOrder.DESC" class="fas fa-sort-down"></i>
                <i *ngIf="sortBy === SortBy.TEAM_NAME && sortOrder === SortOrder.ASC" class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortCoursesBy(SortBy.RESPONDENT_NAME)">
              Name
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i
                  *ngIf="sortBy === SortBy.RESPONDENT_NAME && sortOrder === SortOrder.DESC"
                  class="fas fa-sort-down"
                ></i>
                <i *ngIf="sortBy === SortBy.RESPONDENT_NAME && sortOrder === SortOrder.ASC" class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortCoursesBy(SortBy.RESPONDENT_EMAIL)">
              Email
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i
                  *ngIf="sortBy === SortBy.RESPONDENT_EMAIL && sortOrder === SortOrder.DESC"
                  class="fas fa-sort-down"
                ></i>
                <i *ngIf="sortBy === SortBy.RESPONDENT_EMAIL && sortOrder === SortOrder.ASC" class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortCoursesBy(SortBy.SESSION_END_DATE)">
              Current Deadline
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i
                  *ngIf="sortBy === SortBy.SESSION_END_DATE && sortOrder === SortOrder.DESC"
                  class="fas fa-sort-down"
                ></i>
                <i *ngIf="sortBy === SortBy.SESSION_END_DATE && sortOrder === SortOrder.ASC" class="fas fa-sort-up"></i>
              </span>
            </th>
          </tr>
        </thead>
        <tr
          *ngFor="let student of studentsOfCourse; index as i"
          [ngClass]="{ 'table-warning': student.hasExtension === true }"
        >
          <td class="align-center" (click)="selectStudent(i)">
            <label><input type="checkbox" [checked]="student.selected"/></label>
          </td>
          <td>{{ student.sectionName }}</td>
          <td>{{ student.teamName }}</td>
          <td>{{ student.studentName }}</td>
          <td>{{ student.studentEmail }}</td>
          <td>{{ student.studentExtensionDeadline | date:'d MMM yyyy hh:mm' }}</td>
        </tr>
      </table>
    </div>
  </div>
</div>
