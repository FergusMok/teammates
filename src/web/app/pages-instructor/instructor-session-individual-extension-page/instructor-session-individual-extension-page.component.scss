.background-color-medium-gray {
  background-color: #E5E5E6;
}

.remind-all {
  width: 250px;
}

.clickable {
  cursor: pointer;
}

.margin-right-5px {
  margin-right: 5px;
}

.card-plain {
  background-color: #EAEFF5;
  background-image: none;
  border: 0;
  box-shadow: 0 0 0 0;
}

.close-header {
  padding-top: .75rem;
  padding-right: 1.25rem;
  margin-bottom: 0;
}

.font-bold {
  font-weight: bold;
}

.border-left-gray {
  border-left: 1px solid #DDD;
}

.margin-top-20px {
  margin-top: 20px;
}

.session-form-buttons > button {
  margin-left: var(--btn-margin);
}

.align-row {
  display: flex;
  flex-direction: row;
  gap: 20px;
}
