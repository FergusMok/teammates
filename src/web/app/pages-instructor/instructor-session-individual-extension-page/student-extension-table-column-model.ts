export interface StudentExtensionTableColumnModel {
    sectionName: string;
    teamName: string;
    studentName: string;
    studentEmail: string;
    studentExtensionDeadline: number;
    hasExtension: boolean;
    selected: boolean;
  }
