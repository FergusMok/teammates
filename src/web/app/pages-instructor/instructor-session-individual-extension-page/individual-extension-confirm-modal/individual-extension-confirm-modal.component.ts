import { Component, EventEmitter, Input, Output } from '@angular/core';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import { TableComparatorService } from 'src/web/services/table-comparator.service';
import { SortBy, SortOrder } from 'src/web/types/sort-properties';
import { StudentExtensionTableColumnModel } from '../instructor-session-individual-extension-page.component';
/**
 * Modal to confirm permanent deletion of a feedback session.
 */
@Component({
  selector: 'individual-extension-confirm-modal',
  templateUrl: './individual-extension-confirm-modal.component.html',
  styleUrls: ['./individual-extension-confirm-modal.component.scss'],
})
export class IndividualExtensionConfirmModalComponent {

  @Input()
  courseId: string = '';

  @Input()
  studentsToExtend: StudentExtensionTableColumnModel[] = [];

  @Input()
  extensionTimestamp: number = 0;

  @Input()
  feedbackSessionName: string = '';

  @Output()
  onConfirmExtensionCallBack: EventEmitter<boolean> = new EventEmitter();

  constructor(public activeModal: NgbActiveModal,
              private tableComparatorService: TableComparatorService) { }

  SortBy: typeof SortBy = SortBy;
  SortOrder: typeof SortOrder = SortOrder;
  sortBy: SortBy = SortBy.SECTION_NAME;
  sortOrder: SortOrder = SortOrder.DESC;
  
  isNotifyStudents: boolean = false;

  onExtend() {
    this.onConfirmExtensionCallBack.emit(this.isNotifyStudents);
  }
  
  sortCoursesBy(by: SortBy): void {
    this.sortBy = by;
    this.sortOrder = this.sortOrder === SortOrder.DESC ? SortOrder.ASC : SortOrder.DESC;
    this.studentsToExtend.sort(this.sortPanelsBy(by));
  }

  sortPanelsBy(by: SortBy): (a: StudentExtensionTableColumnModel, b: StudentExtensionTableColumnModel) => number {
    return (a: StudentExtensionTableColumnModel, b: StudentExtensionTableColumnModel): number => {
      let strA: string;
      let strB: string;
      switch (by) {
        case SortBy.SECTION_NAME:
          strA = a.sectionName;
          strB = b.sectionName;
          break;
        case SortBy.TEAM_NAME:
          strA = a.sectionName;
          strB = b.sectionName;
          break;
        case SortBy.RESPONDENT_NAME:
          strA = a.studentName;
          strB = b.studentName;
          break;
        case SortBy.RESPONDENT_EMAIL:
          strA = a.studentEmail;
          strB = b.studentEmail;
          break;
        //TODO: Session End_Date
        case SortBy.SESSION_END_DATE:
          strA = this.extensionTimestamp.toString();
          strB = this.extensionTimestamp.toString();
          break;
        default:
          strA = "";
          strB = "";
      }
      return this.tableComparatorService.compare(by, this.sortOrder, strA, strB);
    };
  }
}