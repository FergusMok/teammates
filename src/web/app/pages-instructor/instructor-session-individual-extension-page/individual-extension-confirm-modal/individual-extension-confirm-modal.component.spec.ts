import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import {
  FeedbackSession,
  FeedbackSessionPublishStatus, FeedbackSessionSubmissionStatus, ResponseVisibleSetting, SessionVisibleSetting,
} from '../../../../types/api-output';
import { InstructorSessionIndividualExtensionPageModule } from '../instructor-session-individual-extension-page.module';
import { StudentExtensionTableColumnModel } from '../student-extension-table-column-model';
import { IndividualExtensionConfirmModalComponent } from './individual-extension-confirm-modal.component';

describe('IndividualExtensionConfirmModalComponent', () => {
    const testFeedbackSession: FeedbackSession = {
        courseId: 'testId1',
        timeZone: 'Asia/Singapore',
        feedbackSessionName: 'Test Session',
        instructions: 'Instructions',
        submissionStartTimestamp: 1000000000000,
        submissionEndTimestamp: 1500000000000,
        gracePeriod: 0,
        sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
        responseVisibleSetting: ResponseVisibleSetting.AT_VISIBLE,
        submissionStatus: FeedbackSessionSubmissionStatus.OPEN,
        publishStatus: FeedbackSessionPublishStatus.PUBLISHED,
        isClosingEmailEnabled: true,
        isPublishedEmailEnabled: true,
        createdAtTimestamp: 0,
    };

    const studentModel1: StudentExtensionTableColumnModel = {
        sectionName: 'Test Section 1',
        teamName: 'Test Section 1',
        studentName: 'Test Student 1',
        studentEmail: 'testStudent1@gmail.com',
        studentExtensionDeadline: 1500000000000,
        hasExtension: false,
        selected: true,
    };

    const studentModel2: StudentExtensionTableColumnModel = {
        sectionName: 'Test Section 2',
        teamName: 'Test Section 2',
        studentName: 'Test Student 2',
        studentEmail: 'testStudent2@gmail.com',
        studentExtensionDeadline: 1510000000000,
        hasExtension: true,
        selected: true,
    };

    const studentModel3: StudentExtensionTableColumnModel = {
        sectionName: 'Test Section 3',
        teamName: 'Test Section 3',
        studentName: 'Test Student 3',
        studentEmail: 'testStudent3@gmail.com',
        studentExtensionDeadline: 1510000000000,
        hasExtension: true,
        selected: true,
    };

    let component: IndividualExtensionConfirmModalComponent;
    let fixture: ComponentFixture<IndividualExtensionConfirmModalComponent>;

    beforeEach(waitForAsync(() => {
        TestBed.configureTestingModule({
          imports: [
            HttpClientTestingModule,
            InstructorSessionIndividualExtensionPageModule,
          ],
          providers: [NgbActiveModal],
        })
        .compileComponents();
      }));

    beforeEach(() => {
        fixture = TestBed.createComponent(IndividualExtensionConfirmModalComponent);
        component = fixture.componentInstance;
        fixture.detectChanges();
      });

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it('should snap with default fields', () => {
        expect(component).toBeTruthy();
    });

    it('should snap with the extended students', () => {
        component.studentsSelected = [studentModel1, studentModel2, studentModel3];
        component.extensionTimestamp = testFeedbackSession.submissionEndTimestamp;
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();
      });
});
