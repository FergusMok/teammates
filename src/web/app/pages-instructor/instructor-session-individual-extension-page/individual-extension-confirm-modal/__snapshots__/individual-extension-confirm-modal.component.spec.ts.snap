// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndividualExtensionConfirmModalComponent should snap with the extended students 1`] = `
<tm-individual-extension-confirm-modal
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  extensionTimestamp={[Function Number]}
  isNotifyStudents="false"
  onConfirmExtensionCallBack={[Function EventEmitter_]}
  sortBy={[Function Number]}
  sortOrder="0"
  studentsToExtend={[Function Array]}
  tableComparatorService={[Function TableComparatorService]}
>
  <div
    class="modal-header alert-primary"
  >
    <h5
      class="modal-title"
    >
      Confirm extension of feedback session
    </h5>
    <button
      class="close"
      type="button"
    >
      <i
        class="fas fa-times"
      />
    </button>
  </div><div
    class="modal-body"
  >
     The new deadline of 
    <b>
      14 Jul 2017 10:40
    </b>
     will be applied to the following 
    <b>
      3 students
    </b>
     upon confirmation: 
    <div
      class="container ml-1"
    />
    <div
      class="modal-body"
    >
      <div
        class="table-responsive"
      >
        <table
          class="table table-bordered"
          id="student-list-table"
        >
          <thead>
            <tr>
              <th
                class="sortable-header"
              >
                 Section 
                <span
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                  <i
                    class="fas fa-sort-down"
                  />
                </span>
              </th>
              <th
                class="sortable-header"
              >
                 Team 
                <span
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </th>
              <th
                class="sortable-header"
              >
                 Name 
                <span
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </th>
              <th
                class="sortable-header"
              >
                 Email 
                <span
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </th>
              <th
                class="sortable-header"
              >
                 Original Deadline 
                <span
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </th>
            </tr>
          </thead>
          <tr>
            <td>
              Test Section 1
            </td>
            <td>
              Test Section 1
            </td>
            <td>
              Test Student 1
            </td>
            <td>
              testStudent1@gmail.com
            </td>
            <td>
              14 Jul 2017 10:40
            </td>
          </tr>
          <tr>
            <td>
              Test Section 2
            </td>
            <td>
              Test Section 2
            </td>
            <td>
              Test Student 2
            </td>
            <td>
              testStudent2@gmail.com
            </td>
            <td>
              7 Nov 2017 04:26
            </td>
          </tr>
          <tr>
            <td>
              Test Section 3
            </td>
            <td>
              Test Section 3
            </td>
            <td>
              Test Student 3
            </td>
            <td>
              testStudent3@gmail.com
            </td>
            <td>
              7 Nov 2017 04:26
            </td>
          </tr>
        </table>
      </div>
    </div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input ng-untouched ng-pristine ng-valid"
        id="flexCheckChecked"
        type="checkbox"
      />
      <label
        class="form-check-label"
        for="flexCheckChecked"
      >
         Notify students by email about this extension 
      </label>
    </div>
  </div><br /><div
    class="modal-footer"
  >
    <button
      class="btn btn-light"
      type="button"
    >
      No, Cancel the Operation
    </button>
    <button
      class="btn btn-success modal-btn-ok"
      type="button"
    >
      Confirm
    </button>
  </div>
</tm-individual-extension-confirm-modal>
`;
